lib/class-wp-rest-url-details-controller.php

```diff
@@ -0,0 +1,122 @@
+<?php
+/**
+ *
+ *
+ * @package gutenberg
+ * @since 5.?.0
+ */
+
+/**
+ * Controller which provides REST endpoint for retrieving information
+ * from a remote site's HTML response.
+ *
+ * @since 5.?.0
+ *
+ * @see WP_REST_Controller
+ */
+class WP_REST_URL_Details_Controller extends WP_REST_Controller {
+
+	/**
+	 * Constructs the controller.
+	 *
+	 * @access public
+	 */
+	public function __construct() {
+		$this->namespace = '__experimental';
+		$this->rest_base = 'url-details';
+	}
+
+	/**
+	 * Registers the necessary REST API routes.
+	 *
+	 * @access public
+	 */
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base . '/title',
+			array(
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_title' ),
+					'args' => array(
+						'url' => array(
+							'validate_callback' => 'wp_http_validate_url',
+							'sanitize_callback' => 'esc_url_raw',
+						),
+					),
+					'permission_callback' => array( $this, 'get_remote_url_permissions_check' ),
+				),
+			)
+		);
+	}
+
+	/**
+	 * Retrieves the contents of the <title> tag from the HTML
+	 * response.
+	 *
+	 * @access public
+	 * @param WP_REST_REQUEST $request Full details about the request.
+	 * @return String|WP_Error The title text or an error.
+	 */
+	public function get_title( $request ) {
+		$url   = $request->get_param( 'url' );
+		$title = $this->get_remote_url_title( $url );
+
+		return rest_ensure_response( $title );
+	}
+
+	/**
+	 * Checks whether a given request has permission to read remote urls.
+	 *
+	 * @return WP_Error|bool True if the request has access, WP_Error object otherwise.
+	 */
+	public function get_remote_url_permissions_check() {
+		if ( ! current_user_can( 'edit_posts' ) ) {
```

Comment:
This breaks things for custom post types. I'd suggest taking a look at the permissions check in the Themes controller.

Reply:
I look at [the Themes Controller here](https://github.com/WordPress/wordpress-develop/blob/37f8276a18633c8b465211469d77a2dddc0a7bc1/src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php#L60) and updated the code to match.